### PR TITLE
refactor compiler a little bit

### PR DESCRIPTION
**Pull request by mccanne**
This commit refactors the compiler and lake so the compiler can import
the lake and do more sophisticated lake things as we integrate warehouse
style optimizations in the Zed lake.  This reverses the previous pattern
where op.DataAdaptor was used by the compiler to talk to the lake.

For now, op.DataAdaptor still exists but we will get rid of this in a
subsequent PR and localize the lakey-ness disposition into the compiler.

### CODE REVIEW

File: ztest/ztest.go [hunk 1]
```diff
@@ -504,7 +505,7 @@ func runzq(path, zedProgram, input string, outputFlags []string, inputFlags []st
 		// tests.
 		return outbuf.String(), errbuf.String(), err
 	}
-	proc, err := compiler.ParseOp(zedProgram)
+	proc, err := compiler.Parse(zedProgram)
```
Comment (nwt) [inline ztest/ztest.go:508]: Just do this so we don't need to thread the compiler down through all these calls (or even make it part of the ztest interface)?
```suggestion
	proc, err := zruntime.NewCompiler().Parse(zedProgram)
```

File: service/core.go [hunk 2]
```diff
@@ -51,6 +53,7 @@ type Config struct {
 type Core struct {
 	auth            *Auth0Authenticator
 	conf            Config
+	compiler        runtime.Compiler
```
Comment (nwt) [inline service/core.go:56]:
```suggestion
	compiler        runtime.Compiler
	conf            Config
```

File: service/core.go [hunk 3]
```diff
@@ -133,6 +136,7 @@ func NewCore(ctx context.Context, conf Config) (*Core, error) {
 	c := &Core{
 		auth:          authenticator,
 		conf:          conf,
+		compiler:      compiler.NewLakeCompiler(root),
```
Comment (nwt) [inline service/core.go:139]:
```suggestion
		compiler:      compiler.NewLakeCompiler(root),
		conf:          conf,
```

File: compiler/job.go [hunk 4]
```diff
@@ -128,93 +127,74 @@ func isSequentialWithLeadingFrom(o ast.Op) bool {
...
-	return r.optimizer.Parallelize(n)
+func (j *Job) Parallelize(n int) error {
+	return j.optimizer.Parallelize(n)
 }
 
-// ParseOp concatenates the source files in filenames followed by src and parses
+// Parse concatenates the source files in filenames followed by src and parses
 // the resulting program.
-func ParseOp(src string, filenames ...string) (ast.Op, error) {
+func (*anyCompiler) Parse(src string, filenames ...string) (ast.Op, error) {
```
Comment (nwt) [inline compiler/job.go:146]: Nit: Move this down below the `anyCompiler` type declaration.

File: compiler/lake.go [hunk 5]
```diff
@@ -2,34 +2,45 @@ package compiler
 
 import (
 	"errors"
-	"runtime"
+	goruntime "runtime"
```
Comment (nwt) [inline compiler/lake.go:5]: Nit: When one of our packages clashes with the standard library like this, I think it'll be a little less surprising for other readers if we rename our package rather than the the standard library package.

File: lake/index/rule.go [hunk 6]
```diff
@@ -65,10 +64,6 @@ func NewTypeRule(name string, typ zed.Type) *TypeRule {
 }
 
 func NewAggRule(name, prog string) (*AggRule, error) {
-	// make sure it compiles
-	if _, err := compiler.ParseOp(prog); err != nil {
-		return nil, err
-	}
```
Comment (nwt) [inline lake/index/rule.go:71]: Are you sure you want to get rid of this?

If you do, consider removing the error return parameter.

File: runtime/op/sort/sort_test.go [hunk 7]
```diff
@@ -130,7 +132,7 @@ func cat(ss ...string) string {
 	return out
 }
 
-func runTest(t *testing.T, cmd, input, output string) {
+func runTest(t *testing.T, c runtime.Compiler, cmd, input, output string) {
```
Comment (nwt) [inline runtime/op/sort/sort_test.go:135]: This parameter is no longer used, so you can revert the changes to this file.

File: service/handlers.go [hunk 8]
```diff
@@ -38,12 +38,12 @@ func handleQuery(c *Core, w *ResponseWriter, r *Request) {
 	// The client must look at the return code and interpret the result
 	// accordingly and when it sees a ZNG error after underway,
 	// the error should be relay that to the caller/user.
-	query, err := compiler.ParseOp(req.Query)
+	query, err := compiler.Parse(req.Query)
```
Comment (nwt) [inline service/handlers.go:41]:
```suggestion
	query, err := c.compiler.Parse(req.Query)
```

**Review by nwt — APPROVED** [review]

File: lake/index/rule.go (hunk 6 above)
Comment (mccanne) [reply lake/index/rule.go:71]: Ok I put it back.  Needed to pass in the a runtime.Compiler because lake packages can't import compiler.

File: service/handlers.go (hunk 8 above)
Comment (mccanne) [reply service/handlers.go:41]: Hmm, I don't know why this passed our tests because c.compiler has the lake pointer but compiler.Parse() doesn't have a lake.

Comment (mccanne) [reply service/handlers.go]: Oh, figured it out.  Parse() just builds the AST and the lake work is done when building the query from the AST.
